exonum/src/runtime/dispatcher/mod.rs

```diff
@@ -401,57 +452,64 @@ impl Dispatcher {
         &self,
         instance_id: InstanceId,
     ) -> Option<(u32, &dyn Runtime)> {
-        let ServiceInfo { runtime_id, .. } = self.service_infos.get(&instance_id)?;
+        let runtime_id = self
+            .service_infos
+            .get_runtime_for_active_instance(instance_id)?;
         let runtime = self.runtimes[&runtime_id].as_ref();
-        Some((*runtime_id, runtime))
+        Some((runtime_id, runtime))
     }
 
     /// Returns the service matching the specified query.
     pub(crate) fn get_service<'s>(
         &'s self,
-        fork: &Fork,
         id: impl Into<InstanceQuery<'s>>,
     ) -> Option<InstanceDescriptor<'s>> {
-        match id.into() {
-            InstanceQuery::Id(id) => {
-                let name = self.service_infos.get(&id)?.name.as_str();
-                Some(InstanceDescriptor { id, name })
-            }
-
-            InstanceQuery::Name(name) => {
-                // TODO: This may be slow.
-                let id = Schema::new(fork).instances().get(name)?.spec.id;
-                Some(InstanceDescriptor { id, name })
-            }
-        }
+        self.service_infos
+            .get_instance(id)
+            .and_then(|(descriptor, status)| {
+                if status.is_active() {
+                    Some(descriptor)
+                } else {
+                    None
+                }
+            })
     }
 
-    /// Notify the runtimes that it has to shutdown.
+    /// Notifies the runtimes that it has to shutdown.
     pub(crate) fn shutdown(&mut self) {
         for runtime in self.runtimes.values_mut() {
             runtime.shutdown();
         }
     }
 
-    /// Start a previously committed service instance.
-    fn start_service(
+    /// Commits service instance state in the corresponding runtime.
+    fn commit_service_state(
         &mut self,
         snapshot: &dyn Snapshot,
         instance: &InstanceSpec,
+        status: InstanceStatus,
     ) -> Result<(), ExecutionError> {
         // Notify the runtime that the service has been committed.
         let runtime = self
             .runtimes
             .get_mut(&instance.artifact.runtime_id)
             .ok_or(Error::IncorrectRuntime)?;
-        runtime.commit_service(snapshot, instance)?;
+        // TODO Rewrite runtime API to pass status to runtime.
```

Comment:
Will this TODO be solved within this PR? I imaging the runtime should be signaled when its service is shut down.